staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go

```diff
@@ -136,7 +136,7 @@ func newAdmissionMetrics() *AdmissionMetrics {
 
 	// Admission webhook metrics. Each webhook is identified by name.
 	webhook := newMetricSet("webhook",
-		[]string{"name", "type", "operation", "rejected", "namespace"},
+		[]string{"name", "type", "operation", "rejected", "namespace", "code"},
```

Comment:
This will most likely make the cardinality explode for the histogram. Considering that the HTTP code is unbounded for admission webhooks, plus the fact that the histogram has 8 buckets, the number of series will theoretically increase by `8 * inf`. If we were to truncate codes to 600 as in `webhook_rejection_count`, it could theoretically increase by `8 * 600` which is a lot. Even though this shouldn't be the case in practice, it is still worrying.

For these reasons, the HTTP code isn't usually added to the latency histogram, but rather to the request counter. Thus, an idea could be to replace `webhook_rejection_count` by a `webhook_request_total` metric that would account for all HTTP codes. We could also truncate the HTTP code to 600 to bound the cardinality of the metric.

Reply:
Good point.

I agree that `webhook_request_total` would be a more useful metric if we were starting from scratch, but for backward compatibility purposed it's probably best to leave `webhook_rejection_count` as is and instead bound the cardinality of the response codes.

Including the code in the latency metric can be useful for identifying if latency spikes are correlated to a specific error.

Reply:
Even if we were to introduce `webhook_request_total` now, it would generate fewer series than adding the HTTP code to the histogram.

`webhook_rejection_count` is still an ALPHA metric, so we can change it anytime. I am not too familiar with admission webhooks, but shouldn't `webhook_request_total` covers the purpose of having `webhook_rejection_count` and handle your fail open use case at the same time, or are rejections handled differently and can't be deduced from the HTTP code only?

I would agree with the statement that the code can be useful in latency metrics, but I don't think it is worth the cardinality spike in this particular use case, but again I am not too familiar with admission webhooks.
Generally, requests failing with code 4xx or 5xx are faster since they are rejected sooner in the normal code path, so if we only care about the 90th and 99th percentile of request latency, for example, to determine SLOs, then the code would not be really useful. Moreover, when determining SLOs and creating alerts on burn rate, what we really care about are the requests with the highest latency (99th percentile) and the ratio of error requests which can be captured by a `request_total` metric.

Reply:
`webhook_rejection_count` is already being used in a number of places internally at google, so I think modifying it is a non-starter. webhook rejections can be deduced from the HTTP code since they are [converted to >=400](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/errors/statuserror.go#L37), so I'm not entirely sure why the rejection count metric was chosen in the first place.

That's a good point about failing requests being unlikely to contribute to tail latencies. If the cardinality spike is not worth it, I think the best course of action is to add a new `webhook_request_total` metric leaving the other two metrics as is.

I'll update the PR with the changes hopefully in the next day or two.

Reply:
@logicalhan mentioned in yesterday's sig-instrumentation triage meeting that the cardinality spike might not be too problematic since we have some mechanisms to prevent cardinality explosions in k/k. That said, I am still on the fence regarding adding code to the histogram since it might not be very useful to the latency, but happy to discuss it :)

Reply:
@logicalhan seemed to be of the mind that adding the new metric was best when I chatted with him. Since that fits our use case just fine I don't think there's much of a need to modify the latency metric at present.

Reply:
:+1:. Would it make sense to work on the deprecation of `webhook_rejection_count` in favor of `webhook_request_total` in a follow-up effort?

Reply:
This would require a fair amount of work internally. @logicalhan has more context here so I'll defer the decision to him.